staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -0,0 +1,173 @@
+	// But if somehow we get any compilation errors, we track them and then surface them as validation errors.
+	compilationErr error
+
+	// isResourceRoot is true if this validator node is for the root of a resource. Either the root of the
+	// custom resource being validated, or the root of an XEmbeddedResource object.
+	isResourceRoot bool
+}
+
+// NewValidator returns compiles all the CEL programs defined in x-kubernetes-validations extensions
+// of the Structural schema and returns a custom resource validator.
+func NewValidator(s *schema.Structural) Validator {
+	return validator(s, true)
+}
+
+func validator(s *schema.Structural, isResourceRoot bool) Validator {
+	compiledRules, err := Compile(s, isResourceRoot)
+	result := Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot}
+	if s.Items != nil {
+		compiledItem := validator(s.Items, s.Items.XEmbeddedResource)
+		result.Items = &compiledItem
+	}
+	if len(s.Properties) > 0 {
+		result.Properties = make(map[string]Validator, len(s.Properties))
+		for k, prop := range s.Properties {
+			compiledProp := validator(&prop, prop.XEmbeddedResource)
+			result.Properties[k] = compiledProp
+		}
+	}
+	if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+		compiledProp := validator(s.AdditionalProperties.Structural, s.AdditionalProperties.Structural.XEmbeddedResource)
+		result.AdditionalProperties = &compiledProp
+	}
+
+	return result
+}
+
+// Validate validates all x-kubernetes-validations rules in Validator against obj and returns any errors.
+func (s *Validator) Validate(fldPath *field.Path, sts *schema.Structural, obj interface{}) field.ErrorList {
+	if s == nil || obj == nil {
+		return nil
+	}
+
+	errs := s.validateExpressions(fldPath, sts, obj)
+	switch obj := obj.(type) {
+	case []interface{}:
+		return append(errs, s.validateArray(fldPath, sts, obj)...)
+	case map[string]interface{}:
+		return append(errs, s.validateMap(fldPath, sts, obj)...)
+	}
+	return errs
+}
+
+func (s *Validator) validateExpressions(fldPath *field.Path, sts *schema.Structural, obj interface{}) (errs field.ErrorList) {
+	if s.isResourceRoot {
+		sts = model.WithTypeAndObjectMeta(sts)
+	}
+	activation := &validationActivation{obj: obj, structural: sts}
+	if s.compilationErr != nil {
+		errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("failed to compile rules due to error %v", s.compilationErr)))
+		return errs
+	}
+	for i, compiled := range s.compiledRules {
+		rule := sts.XValidations[i]
+		if compiled.Error != nil {
+			errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("failed to compile rule '%s' due to error %v", rule.Rule, compiled.Error)))
+			continue
+		}
+		evalResult, _, err := compiled.Program.Eval(activation)
+		if err != nil {
+			errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("failed to execute rule '%s' due to error %v", rule.Rule, err)))
```

Comment:
is this the path that runtime errors in the cel expression trigger? are there expected errors like `no_such_field` that we should report more concisely?

Reply:
`rule runtime error: {Rule}` is probably sufficient. I'll go with that. I've updated the other error descriptions too.

Reply:
I've reviewed the possible runtime errors, made most more concise, and made the "no such overload" one quite a bit more verbose since I don't expect most kubernetes users to know what that means.